pkg/frontend/querymiddleware/roundtrip.go

```diff
@@ -75,9 +72,11 @@ func (cfg *Config) Validate() error {
 		if err := cfg.ResultsCacheConfig.Validate(); err != nil {
 			return errors.Wrap(err, "invalid ResultsCache config")
 		}
-	} else {
-		if cfg.MaxSeriesPerShard > 0 {
-			return fmt.Errorf("-%s may only be enabled in conjunction with -%s", maxSeriesPerShardFlagName, cacheResultsFlagName)
+	}
+
+	if cfg.cardinalityBasedShardingEnabled() {
```

Comment:
```suggestion
	if if cfg.CacheResults || cfg.cardinalityBasedShardingEnabled() {
```

Reply:
fixed in https://github.com/grafana/mimir/pull/4254/commits/94de67f4f6ba693a6cc7766183f2b03a5c4b5fae